File: app/src/main/java/com/alexkn/syntact/hangwords/ui/util/DiffCallback.java
package com.alexkn.syntact.hangwords.ui.util;

import com.alexkn.syntact.hangwords.util.Identifiable;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;

public class DiffCallback<T extends Identifiable> extends DiffUtil
        .ItemCallback<T> {

    @Override
    public boolean areItemsTheSame(@NonNull T oldItem, @NonNull T newItem) {
        return oldItem.getId().equals(newItem.getId());
    }

    @Override
    public boolean areContentsTheSame(@NonNull T oldItem, @NonNull
            T newItem) {
        return oldItem.equals(newItem);
    }
}

File: app/src/main/java/dev/alexknittel/syntact/presentation/common/DiffCallback.kt
package dev.alexknittel.syntact.presentation.common

import android.annotation.SuppressLint
import androidx.recyclerview.widget.DiffUtil
import dev.alexknittel.syntact.core.model.Identifiable

class DiffCallback<T : Identifiable<*>> : DiffUtil.ItemCallback<T>() {

    override fun areItemsTheSame(oldItem: T, newItem: T): Boolean {
        return oldItem.id == newItem.id
    }

    @SuppressLint("DiffUtilEquals")
    override fun areContentsTheSame(oldItem: T, newItem: T): Boolean {
        return oldItem == newItem
    }
}
